add `slice::array_chunks` to std

lcnr opened the pull request:
Now that #74113 has landed, these methods are suddenly usable. A rebirth of #72334

Tests are directly copied from `chunks_exact` and some additional tests for type inference.

r? @withoutboats as you are both part of t-libs and working on const generics. closes #60735

lcnr committed:
improve chunks + windows err on size 0

lcnr committed:
adds `slice::array_chunks`

lcnr committed:
add tests for array_chunks

lcnr committed:
liballoc export ArrayChunks

lcnr commented on src/libcore/tests/slice.rs:
```diff
@@ -473,6 +473,97 @@ fn test_chunks_exact_mut_zip() {
     assert_eq!(v1, [13, 14, 19, 20, 4]);
 }
 
+#[test]
+fn test_array_chunks_infer() {
+    let v: &[i32] = &[0, 1, 2, 3, 4, -4];
+    let c = v.array_chunks();
+    for &[a, b, c] in c {
+        assert_eq!(a + b + c, 3);
+    }
+
+    let v2: &[i32] = &[0, 1, 2, 3, 4, 5, 6];
+    let total = v2.array_chunks().map(|&[a, b]| a * b).sum::<i32>();
+    assert_eq!(total, 2 * 3 + 4 * 5);
+}
```
Look at this beauty

jplatte commented:
Is there also a PR for `array_windows`? :)

lcnr commented:
Feel free to open one once `T-libs` decides this gets merged.
There are too many methods for me to do this all in one go :smile:

DutchGhost commented:
Neat!! Just one little thought, should `.array_chunks_mut::<N>()` also be included in this PR?

varkor commented:
Presumably this fixes https://github.com/rust-lang/rust/issues/60735.

withoutboats commented:
This looks good to me. 

Only question to me is the name. Too bad that this is just a strictly better version of the existing chunks_exact. array_chunks is good, but maybe something that starts with chunks_ to group it with the other chunk methods alphabetically (chunks_array doesn't sound amazing but I don't have an immediate better alternative).

lcnr commented:
There are rare edge cases where `chunks_exact` is useful (when dealing with variably sized chunks) :sweat_smile:

The docs order is a valid concern I didn't previously consider.
However, once we add the remaining const generic methods:
- `array_chunks_mut`
- `array_windows`
- `array_rchunks`
- `array_rchunks_mut`

We would end up with all const generics methods grouped together which IMO is also desirable.

I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.

leonardo-m commented on src/libcore/slice/mod.rs:
```diff
@@ -841,6 +841,41 @@ impl<T> [T] {
+    ///
+    /// # Panics
+    ///
+    /// Panics if `N` is 0.
```
Could you please add a note here that says that future versions of this are allowed to panic at compile-time if `N` is 0 instead of run-time?

lcnr replied:
While that's desirable, I am not sure how feasible it is to change this after stabilization.

Opened a discussion on zulip about this: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/array_chunks.20of.20size.200/near/204001667

lcnr replied:
Ended up changing `array_chunks::<0>()` to a compile time error. While we may removing this bound before this method hits stable, it seems like the better default for now.
```
error[E0277]: the trait bound `[(); 0]: core::slice::sealed::NonZero` is not satisfied
```

cuviper replied:
FWIW, there's also an upper bound on array lengths for any non-ZST `T`, but that should get a compile-time ``error: the type `...` is too big for the current architecture``.

lcnr replied:
Aaah, nm. Adding this bound breaks type inference. Looks like we may have to live with a runtime error for now

leonardo-m commented:
I've just seen that in my codebase I can replace all windows with array_windows.

cuviper commented on src/libcore/slice/mod.rs:
```diff
@@ -5419,6 +5453,151 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksExactMut<'a, T> {
+#[derive(Debug)]
+#[unstable(feature = "array_chunks", issue = "none")]
+pub struct ArrayChunks<'a, T: 'a, const N: usize> {
+    v: &'a [T],
```
Another possibility is to cast `&[T]` to `&[[T; N]]` up front, then store `Iter<'a, [T; N]>` here. That will at least simplify the `Iterator` implementation, making it totally safe.

lcnr replied:
That's a good idea! Thanks :heart:

lcnr replied:
Can't derive `Debug` in this case. Will add this as a `FIXME` as it is otherwise a lot cleaner and update it once #74060 lands

jplatte commented on src/test/ui/const-generics/std/array-chunks-zero.stderr:
```diff
@@ -0,0 +1,21 @@
+LL |     slice.array_chunks::<0>();
+   |           ^^^^^^^^^^^^ the trait `core::slice::sealed::NonZero` is not implemented for `[(); 0]`
+   |
+   = help: the following implementations were found:
+             <[(); 0] as core::slice::sealed::NonZero>
```
That's one unhelpful hint 😅

lcnr replied:
opened #74382

SimonSapin commented:
Since this has the semantics of `chunks_exact` and *not* those of `chunks`, I feel this should be called `array_chunks_exact` (even though that’s lengthy and the corresponding `array_chunks` with `chunks` semantics is impossible).

lcnr commented:
> Since this has the semantics of chunks_exact and not those of chunks, I feel this should be called array_chunks_exact

I personally do not feel this way as `array_` very strongly implies fixed length. `[T]` would be a slice.
Using `exact` here would end up being redundant in my view.

snawaz commented:
> The docs order is a valid concern I didn't previously consider.
> However, once we add the remaining const generic methods:
> 
> * `array_chunks_mut`
> * `array_windows`
> * `array_rchunks`
> * `array_rchunks_mut`
> 
> We would end up with all const generics methods grouped together which IMO is also desirable.
> 
> I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.

How about

* `chunks_const`
* `chunks_mut_const`
* `windows_const` 
* `rchunks_const`
* `rchunks_mut_const`

? 

After all, these functions are based on `const` generics and thus  take a `const N` as generic argument.

snawaz commented:
> The docs order is a valid concern I didn't previously consider.
> However, once we add the remaining const generic methods:
> 
> * `array_chunks_mut`
> * `array_windows`
> * `array_rchunks`
> * `array_rchunks_mut`
> 
> We would end up with all const generics methods grouped together which IMO is also desirable.
> 
> I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.

How about

* `chunks_const`
* `chunks_mut_const`
* `windows_const` 
* `rchunks_const`
* `rchunks_mut_const`

? 

After all, these functions are based on `const` generics and thus  take a `const N` as generic argument. I believe this also addresses the @SimonSapin's concern (to use `_exact` in its name), as `_const` implies that anyway &mdash; that too, more strongly! 

Or we could use `_fixed` suffix instead. I'd prefer the `_const` though. 😄

snawaz commented:
> The docs order is a valid concern I didn't previously consider.
> However, once we add the remaining const generic methods:
> 
> * `array_chunks_mut`
> * `array_windows`
> * `array_rchunks`
> * `array_rchunks_mut`
> 
> We would end up with all const generics methods grouped together which IMO is also desirable.
> 
> I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.

How about

* `chunks_const`
* `chunks_mut_const`
* `windows_const` 
* `rchunks_const`
* `rchunks_mut_const`

? 

After all, these functions are based on `const` generics and thus  take a `const N` as generic argument. I believe this also addresses @SimonSapin's concern (to use `_exact` in its name), as `_const` implies that anyway &mdash; that too, more strongly! 

Or we could use `_fixed` suffix instead. I'd prefer the `_const` though. 😄

snawaz commented:
@lcnr 

> The docs order is a valid concern I didn't previously consider.
> However, once we add the remaining const generic methods:
> 
> * `array_chunks_mut`
> * `array_windows`
> * `array_rchunks`
> * `array_rchunks_mut`
> 
> We would end up with all const generics methods grouped together which IMO is also desirable.
> 
> I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.

How about

* `chunks_const`
* `chunks_mut_const`
* `windows_const` 
* `rchunks_const`
* `rchunks_mut_const`

? 

After all, these functions are based on `const` generics and thus  take a `const N` as generic argument. I believe this also addresses @SimonSapin's concern (to use `_exact` in its name), as `_const` implies that anyway &mdash; that too, more strongly! 

Or we could use `_fixed` suffix instead. I'd prefer the `_const` though. 😄

DutchGhost commented:
> @lcnr
> 
> > The docs order is a valid concern I didn't previously consider.
> > However, once we add the remaining const generic methods:
> > 
> > * `array_chunks_mut`
> > * `array_windows`
> > * `array_rchunks`
> > * `array_rchunks_mut`
> > 
> > We would end up with all const generics methods grouped together which IMO is also desirable.
> > I fairly strongly prefer `array_chunks` over `chunks_array` and wasn't yet able to come up with a good alternative starting with `chunks_`. `chunks_const` doesn't sound completely horrible, but I still prefer `array_chunks` here.
> 
> How about
> 
>     * `chunks_const`
> 
>     * `chunks_mut_const`
> 
>     * `windows_const`
> 
>     * `rchunks_const`
> 
>     * `rchunks_mut_const`
> 
> 
> ?
> 
> After all, these functions are based on `const` generics and thus take a `const N` as generic argument. I believe this also addresses @SimonSapin's concern (to use `_exact` in its name), as `_const` implies that anyway — that too, more strongly!
> 
> Or we could use `_fixed` suffix instead. I'd prefer the `_const` though. smile

I'd prefer something else than `_const`, as it sort of almost suggest that the function is a `const fn` and that the resulting Iterator is `const fn` compatible

cuviper commented:
`mut_const` is too much of a contradiction, even though those qualifiers are referring to different things.

snawaz commented:
> `mut_const` is too much of a contradiction, even though those qualifiers are referring to different things.

That's true. 😂 

Few more name suggestions along with others discussed so far, just for comparison, all at once:

- `array_chunks` (original; will not group with the old members)
- `array_chunks_mut` 

vs

- `chunks_array` (flipped; but does not look good)
- `chunks_mut_array` 

vs

- `chunks_const` 
- `chunks_mut_const` (looks contradictory)

vs

- `chunks_fixed` (does not look good enough to me)
- `chunks_mut_fixed`

vs

- `chunks_static` (as the size is known statically)
- `chunks_mut_static`

Can't think of any other suffix 🤔

Hope the last one works for @lcnr  and others.

lcnr commented:
I personally still prefer `array_chunks` here as I somewhat elaborated on in `https://github.com/rust-lang/rust/pull/74373#issuecomment-658914061`.

At least for me personally `static` is too strongly related to lifetimes to also consider it to describe sizes.
I do think `chunks_fixed` seems okay, so I would also be fine with that.

snawaz commented:
Fair enough. 

After considering all alternative names I could think of, `array_chunks` seems pretty good to me now, _also_ from another perspective: it could return `&[T; N]` instead of `&[T]`. And since each element (chunk) is an array, it could be named as `chunks_as_array` &mdash; looks lengthy though, but it (or simply `chunks_array`) has one advantage: grouping all _chunks_ functions.

cuviper commented:
A slight variation could use the adjective -- `arrayed_chunks` or `chunks_arrayed`.

> it could return `&[T; N]` instead of `&[T]`.

That's exactly what this PR does! The closures like `|[a, b]|` would have refutable patterns if they were slices.

jyn514 commented on src/libcore/slice/mod.rs:
```diff
@@ -5419,6 +5453,153 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksExactMut<'a, T> {
+/// This struct is created by the [`array_chunks`] method on [slices].
+///
+/// [`array_chunks`]: ../../std/primitive.slice.html#method.array_chunks
+/// [`remainder`]: ../../std/slice/struct.ArrayChunks.html#method.remainder
+/// [slices]: ../../std/primitive.slice.html
```
These links will be broken if viewed from the `core` docs. Unfortunately `slice` can't be fixed due to #63351.
```suggestion
/// [`remainder`]: ArrayChunks::remainder
/// [slices]: ../../std/primitive.slice.html
```

pickfire commented on src/libcore/slice/mod.rs:
```diff
@@ -834,13 +834,47 @@ impl<T> [T] {
+    ///
+    /// The chunks are slices and do not overlap. If `N` does not divide the length of the
+    /// slice, then the last up to `N-1` elements will be omitted and can be retrieved
+    /// from the `remainder` function of the iterator.
```
Could we mention that this is equivalent to `chunks`? Not specifying this might let people get the wrong idea, `chunks_exact`. Maybe we should also have a version for that?

lcnr replied:
This is actually equivalent to `chunks_exact`, as it returns `&[T; N]`, which must always contain exactly N elements.

Because of this, implementing `chunks` usings arrays/const generics doesn't really make sense here.

pickfire replied:
Oh, then how about a version for `chunks`? If there is a generic version for `chunks_exact` I think there should be one for `chunks` too.

If we took `chunks_array` then the `chunks` version will be `chunks_array_not_exact`? Or maybe we should keep the `_exact` for this?

lcnr committed:
add note to array_chunks

withoutboats commented:
I found the argument that the array returning iterator methods will be grouped together pretty compelling. I also don't think there's reason to make the name longer than it is (its not like its possible to have a non-exact array_chunks), especially since it'll often be combined with method chaining adapters. I'm happy with `array_chunks` personally.

So for now:

@rfcbot fcp merge

withoutboats commented:
@rfcbot fcp merge

wasn't tagged libs

pickfire commented:
> I found the argument that the array returning iterator methods will be grouped together pretty compelling. I also don't think there's reason to make the name longer than it is (its not like its possible to have a non-exact array_chunks), especially since it'll often be combined with method chaining adapters. I'm happy with array_chunks personally.

Ah, I just realized it is not possible to have chunks since we already specified for N items.

Looking at the function naming? I think this may be confusing for existing developers, I thought this is the equivalent of `chunks` at first look and is looking for another version of `chunks_exact` but to my surprise this is the equivalent of `chunks_exact`. I believe the function should contain `_exact` to not surprise existing developers who knew `chunks_exact`.

jethrogb commented:
Are there also plans for `split_at_array` and  `split_at_mut_array`?

pickfire commented:
> Are there also plans for split_at_array and split_at_mut_array?

I don't think it is necessary, I don't see the use of `const N`.

jethrogb commented:
```
impl [T] {
    fn split_at_array<const N: usize>(&self) -> (&[T; N], &[T])
}
```

cuviper commented:
Bike-shedding `split_at_array` -- I think "at" only makes less sense when there's no argument, only a generic parameter that may often be omitted and inferred. So I'd just call it `split_array` and `split_array_mut`.

jplatte commented:
To further add to that (maybe we should move this into a new issue?) what would be really nice would be

```rust
impl<T, const N: usize> [T; N] {
    fn split_array<const M: usize>(&self) -> (&[T; M], &[T; {N - M}])
}
```

but I don't think that works yet. Introducing the slice method without introducing the array method would make adding the array method later a breaking change, right?

cuviper commented:
> Introducing the slice method without introducing the array method would make adding the array method later a breaking change, right?

I think that's technically a minor break, which is allowed. It's the same problem we have for array `into_iter()`, but that one has enough reach that we have real reason to be scared of breaking it.

jethrogb commented:
Now that we know about the forwards-compatibility issue, just add an unstable method with that name on arrays, forcing people to manually convert to a slice,  until const generics is powerful enough.

pickfire commented:
What if users want to split using sort of like match?
```rust
let [first, ..body, last] = v.split_array();
```
```rust
impl<T, const N: usize> [T; N] {
    fn split_array<const M: usize>(&self) -> (&[T; M], &[T; {N - M}], ...)
}
```

DutchGhost commented:
> 
> 
> What if users want to split using sort of like match?
> 
> ```rust
> let [first, ..body, last] = v.split_array();
> ```
> 
> Of course, this isn't possible.
> 
> ```rust
> impl<T, const N: usize> [T; N] {
>     fn split_array<const M: usize>(&self) -> (&[T; M], &[T; {N - M}], ...)
> }
> ```

You can make it work by introducing 2 const parameters,
```Rust
#![feature(bindings_after_at)]
#![feature(const_generics)]

trait Splitter {
    type Item;
    fn split_array<const N: usize, const M: usize>(&self) -> (&[Self::Item; N], &[Self::Item; M]);
}

impl<T, const Q: usize> Splitter for [T; Q] {
    type Item = T;

    fn split_array<const N: usize, const M: usize>(&self) -> (&[Self::Item; N], &[Self::Item; M]) {
        assert!(N + M == Q);

        let (lhs, rhs) = self.split_at(N);

        unsafe {
            let lhs_array = &(*lhs.as_ptr().cast());
            let rhs_array = &(*rhs.as_ptr().cast());

            (lhs_array, rhs_array)
        }
    }
}

fn main() {
    let array = [1, 2, 3, 4, 5];

    match array.split_array() {
        (a1 @ [a, b, c], a2 @ [d, e, f]) => {
            dbg!(a1);
            dbg!(a2);
        }
    }
}
```
this was just very quickly put together. The conversation about this method should probably be moved to another issue though, as @jplatte mentioned earlier.

jplatte commented:
> What if users want to split using sort of like match?
> 
> ```rust
> let [first, ..body, last] = v.split_array();
> ```

Your specific example shows a slice pattern¹, which is different from

```rust
let v = [0; 7];
let (start, middle, end) = v.split_array::<2, 3, 2>();
```

Since with the syntax you wrote `first` and `last` are single elements, while the `split` functions return sub-slices.

Anyway, I've opened a new issue about const-generic array splitting: #74674

¹ pretty much, just the subslice binding has to be changed and [it works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c13c48c43c5c4322d7c97de73deb0f5c)

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

lcnr commented:
Changed the implementation to use an iterator over `[T; N]` instead of manually reimplementing this as recommended by @cuviper.

I also added a note that we will probably concert the runtime panic on `N == 0` into a compile time error before stabilizing this method.

scottmcm commented:
Looks like this accidentally added a submodule?
![image](https://user-images.githubusercontent.com/18526288/89006125-dbb45500-d2ba-11ea-85ea-67afb200678e.png)

lcnr committed:
use Iter<'_, [T; N]> in array_chunks

lcnr commented:
yeah. Thanks :sweat_smile:

withoutboats commented:
@bors r+

bors commented:
:pushpin: Commit 1b90e78393866dbedc859bd09a353ce83112e475 has been approved by `withoutboats`

<!-- @bors r=withoutboats 1b90e78393866dbedc859bd09a353ce83112e475 -->
<!-- homu: {"type":"Approved","sha":"1b90e78393866dbedc859bd09a353ce83112e475","approver":"withoutboats"} -->

lzutao commented:
Is this intentional?
![image](https://user-images.githubusercontent.com/15225902/89091805-b6b50480-d3d6-11ea-87ff-a666036ef353.png)

cuviper commented:
That submodule was noticed before, then came back with the tracking issues...

@bors r-

lcnr committed:
add tracking issue

lcnr commented:
................what? sry

fixed it @bors r=withoutboats

bors commented:
:pushpin: Commit d51b71a35a816f4be56f77d1d1a6f4095352649e has been approved by `withoutboats`

<!-- @bors r=withoutboats d51b71a35a816f4be56f77d1d1a6f4095352649e -->
<!-- homu: {"type":"Approved","sha":"d51b71a35a816f4be56f77d1d1a6f4095352649e","approver":"withoutboats"} -->

bors commented:
:hourglass: Testing commit d51b71a35a816f4be56f77d1d1a6f4095352649e with merge b5eae9c44d713779a08e6db352088d45cad3e9b6...
<!-- homu: {"type":"BuildStarted","head_sha":"d51b71a35a816f4be56f77d1d1a6f4095352649e","merge_sha":"b5eae9c44d713779a08e6db352088d45cad3e9b6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/934848795), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35724)
Approved by: withoutboats
Pushing b5eae9c44d713779a08e6db352088d45cad3e9b6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"withoutboats","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35724","checks-actions":"https://github.com/rust-lang-ci/rust/runs/934848795"},"merge_sha":"b5eae9c44d713779a08e6db352088d45cad3e9b6"} -->
